lib/system/coroutines.nim

```diff
@@ -0,0 +1,125 @@
+## Internal module included by the system module. Implements the basic
+## compiler interface for coroutines, plus a low-level API for working
+## with them.
+
+type
+  CoroutineBase {.compilerproc.} = ref object of RootObj
+    ## The internal base type of all coroutine instance type. Carries the
+    ## state all coroutine instances have.
+    fn: proc(c: CoroutineBase): CoroutineBase {.nimcall, raises: [].}
+      ## internal procedure
+    state: int32
+      ## internal state value
+    running: bool
```

Comment:
Discussed with @zerbina in [matrix chat](https://matrix.to/#/!MAXxvfCYnfYgZqSizw:matrix.org/$hgmWkMbDblFbku-WGiNS_IJbUMUG_3tRSn1SZiWBAO0?via=matrix.org&via=envs.net&via=furs.chat), but we can get rid of `running` by encoding it in `state`.

Any negative `state` value means "not running".

Where resume would do:

```nim
if c.state < -3:
  c.state = (-c.state) + 3
# run...
if  c.state >= 0:
  c.state = (-c.state) - 3
```

Reply:
I've applied my suggestion for removing the extra `running` boolean field. There were some problems with it (the inversion logic was wrong, and -4 has to be used instead of -3), which I addressed.